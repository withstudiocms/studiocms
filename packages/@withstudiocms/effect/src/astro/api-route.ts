import type { APIContext, APIRoute } from 'astro';
import { runEffect } from '../index.js';
import type {
	EffectAPIRouteHandler,
	EffectRouteOptions,
	ExtractDefinedRoutes,
	RouteHandlerConfig,
	RouteHandlers,
} from './types.js';
import { applyCors, validateRequest } from './utils/api-route.js';

/**
 * Defines an API route handler for Astro using an effect-based approach.
 *
 * @param context - The API context provided by Astro for the current request.
 * @returns A function that takes an `EffectAPIRouteHandler`, executes it with the provided context,
 *          and returns a `Promise<Response>` representing the HTTP response.
 *
 * @example
 * ```typescript
 * export const GET = defineAPIRoute(context)(async (ctx) => {
 *   // Your effectful logic here
 *   return new Response("Hello, world!");
 * });
 * ```
 *
 * @deprecated Use `createEffectRoute` instead for better clarity and consistency.
 */
export const defineAPIRoute =
	(context: APIContext) =>
	async (fn: EffectAPIRouteHandler): Promise<Response> =>
		await runEffect(fn(context));

/**
 * Creates an Astro API route handler that executes the provided effectful handler function.
 *
 * @param fn - The effectful API route handler function to be executed. It should accept an `APIContext` and return an effect.
 * @returns An Astro `APIRoute` handler that runs the effect and returns a `Response`.
 */
export const createEffectAPIRoute = (fn: EffectAPIRouteHandler): APIRoute => {
	return async (context: APIContext): Promise<Response> => {
		return await runEffect(fn(context));
	};
};

/**
 * Wraps an API route handler with effectful middleware, including CORS, validation, timeout, and custom hooks.
 *
 * @param fn - The main effectful API route handler function to execute.
 * @param options - Optional configuration for the route, including CORS, validation, timeout, and middleware hooks.
 * @returns An `APIRoute` function compatible with Astro's API routes.
 *
 * @remarks
 * This utility provides a standardized way to handle API routes with common concerns such as:
 * - Applying CORS headers and handling preflight requests.
 * - Running pre-effect middleware (`onBeforeEffect`).
 * - Validating incoming requests (`validate`).
 * - Enforcing a timeout for the effect execution.
 * - Running post-success (`onSuccess`) and error (`onError`) middleware.
 * - Ensuring CORS headers are present on all responses, including errors.
 *
 * @example
 * ```typescript
 * export const POST = withEffect(async (context) => {
 *   // Your effect logic here
 * }, {
 *   cors: { origin: '*' },
 *   validate: { body: z.object({ name: z.string() }) },
 *   timeout: 5000,
 *   onSuccess: async (response, context) => response,
 *   onError: async (error, context) => new Response('Error', { status: 500 }),
 * });
 * ```
 */
export const withEffectAPI = (
	fn: EffectAPIRouteHandler,
	options: EffectRouteOptions = {}
): APIRoute => {
	return async (context: APIContext): Promise<Response> => {
		try {
			// Apply CORS headers if configured
			const corsHeaders = applyCors(context, options.cors);

			// Handle preflight requests
			if (context.request.method === 'OPTIONS') {
				return new Response(null, {
					status: 204,
					headers: corsHeaders,
				});
			}

			// Apply pre-effect middleware
			let processedContext = context;
			if (options.onBeforeEffect) {
				processedContext = await options.onBeforeEffect(context);
			}

			// Validate request if validators provided
			if (options.validate) {
				const validationError = await validateRequest(processedContext, options.validate);
				if (validationError) {
					const errorResponse = new Response(
						JSON.stringify({ error: 'Validation failed', details: validationError }),
						{ status: 400, headers: { 'Content-Type': 'application/json', ...corsHeaders } }
					);
					return errorResponse;
				}
			}

			// Execute the effect with timeout if specified
			let effectPromise = runEffect(fn(processedContext));
			if (options.timeout) {
				effectPromise = Promise.race([
					effectPromise,
					new Promise<never>((_, reject) =>
						setTimeout(() => reject(new Error('Request timeout')), options.timeout)
					),
				]);
			}

			const response = await effectPromise;

			// Add CORS headers to response
			if (corsHeaders && Object.keys(corsHeaders).length > 0) {
				Object.entries(corsHeaders).forEach(([key, value]) => {
					response.headers.set(key, value);
				});
			}

			// Apply success middleware
			return options.onSuccess ? await options.onSuccess(response, processedContext) : response;
		} catch (error) {
			const corsHeaders = applyCors(context, options.cors);

			if (options.onError) {
				const errorResponse = await options.onError(error, context);
				// Ensure CORS headers on error responses too
				Object.entries(corsHeaders).forEach(([key, value]) => {
					errorResponse.headers.set(key, value);
				});
				return errorResponse;
			}

			// Default error handling with CORS
			return new Response(
				JSON.stringify({
					error: error instanceof Error ? error.message : 'Internal Server Error',
				}),
				{
					status: error instanceof Error && error.message === 'Request timeout' ? 408 : 500,
					headers: {
						'Content-Type': 'application/json',
						...corsHeaders,
					},
				}
			);
		}
	};
};

/**
 * Creates API route handlers with effect support for each HTTP method provided.
 *
 * This function takes a set of route handlers and an optional configuration object,
 * then returns an object mapping HTTP methods to their corresponding API route handlers,
 * each wrapped with effect middleware and merged configuration.
 *
 * The return type is precisely typed to only include the methods that were provided
 * in the handlers parameter, eliminating undefined from the union types.
 *
 * @param handlers - An object containing route handlers keyed by HTTP method (e.g., 'GET', 'POST').
 * @param config - Optional global configuration for all handlers. Can include method-specific overrides via the `methods` property.
 * @returns An object mapping each provided HTTP method to its configured API route handler.
 */
export const createEffectAPIRoutes = <T extends RouteHandlers>(
	handlers: T,
	config: RouteHandlerConfig = {}
): ExtractDefinedRoutes<T> => {
	const routes = {} as ExtractDefinedRoutes<T>;

	for (const [method, handler] of Object.entries(handlers)) {
		if (handler) {
			const methodKey = method as keyof RouteHandlers;

			// Merge global config with method-specific overrides
			const methodConfig = {
				...config,
				...(config.methods?.[methodKey] || {}),
			};

			// Remove the methods property from the final config
			const { methods: _droppedMethods, ...finalConfig } = methodConfig;

			// TypeScript knows this is safe because we checked `if (handler)` above
			// biome-ignore lint/suspicious/noExplicitAny: `routes` is typed as `ExtractDefinedRoutes<T>`
			(routes as any)[methodKey] = withEffectAPI(handler as EffectAPIRouteHandler, finalConfig);
		}
	}

	return routes;
};

/**
 * A builder class for defining API route handlers with optional per-method and global configuration.
 *
 * Provides a fluent interface for registering HTTP method handlers (`get`, `post`, `put`, `delete`, `patch`, `all`)
 * and setting global or per-method options. Use `build()` to generate the final route configuration.
 *
 * Each method returns a new builder type that accumulates the added methods, ensuring precise typing
 * in the final built routes object.
 *
 * @example
 * ```typescript
 * const routes = new EffectRouteBuilder()
 *   .withGlobalConfig({ cors: { origin: true } })
 *   .get(getHandler, { timeout: 3000 })
 *   .post(postHandler)
 *   .build();
 * // routes has type: { GET: APIRoute, POST: APIRoute } (no undefined!)
 * ```
 */

// biome-ignore lint/complexity/noBannedTypes: `T` is used to accumulate handlers
export class EffectAPIRouteBuilder<T extends Partial<RouteHandlers> = {}> {
	private handlers: T;
	private config: RouteHandlerConfig = {};

	constructor(handlers: T = {} as T) {
		this.handlers = handlers;
	}

	// Set global configuration
	public withGlobalConfig(config: EffectRouteOptions): EffectAPIRouteBuilder<T> {
		this.config = { ...this.config, ...config };
		return this;
	}

	private addHandler<K extends keyof RouteHandlers, U extends EffectAPIRouteHandler>(
		method: K,
		handler: U,
		options?: Partial<EffectRouteOptions>
	): EffectAPIRouteBuilder<T & Record<K, U>> {
		const newHandlers = { ...this.handlers, [method]: handler } as T & Record<K, U>;
		const newBuilder = new EffectAPIRouteBuilder(newHandlers);
		newBuilder.config = { ...this.config };

		if (options) {
			newBuilder.config.methods = { ...this.config.methods, [method]: options };
		}
		return newBuilder;
	}

	// Add handlers with fluent interface - each method returns a new type that includes the added method
	public get<U extends EffectAPIRouteHandler>(
		handler: U,
		options?: Partial<EffectRouteOptions>
	): EffectAPIRouteBuilder<T & { GET: U }> {
		return this.addHandler('GET', handler, options) as EffectAPIRouteBuilder<T & { GET: U }>;
	}

	public post<U extends EffectAPIRouteHandler>(
		handler: U,
		options?: Partial<EffectRouteOptions>
	): EffectAPIRouteBuilder<T & { POST: U }> {
		return this.addHandler('POST', handler, options) as EffectAPIRouteBuilder<T & { POST: U }>;
	}

	public put<U extends EffectAPIRouteHandler>(
		handler: U,
		options?: Partial<EffectRouteOptions>
	): EffectAPIRouteBuilder<T & { PUT: U }> {
		return this.addHandler('PUT', handler, options) as EffectAPIRouteBuilder<T & { PUT: U }>;
	}

	public delete<U extends EffectAPIRouteHandler>(
		handler: U,
		options?: Partial<EffectRouteOptions>
	): EffectAPIRouteBuilder<T & { DELETE: U }> {
		return this.addHandler('DELETE', handler, options) as EffectAPIRouteBuilder<T & { DELETE: U }>;
	}

	public patch<U extends EffectAPIRouteHandler>(
		handler: U,
		options?: Partial<EffectRouteOptions>
	): EffectAPIRouteBuilder<T & { PATCH: U }> {
		return this.addHandler('PATCH', handler, options) as EffectAPIRouteBuilder<T & { PATCH: U }>;
	}

	public all<U extends EffectAPIRouteHandler>(
		handler: U,
		options?: Partial<EffectRouteOptions>
	): EffectAPIRouteBuilder<T & { ALL: U }> {
		return this.addHandler('ALL', handler, options) as EffectAPIRouteBuilder<T & { ALL: U }>;
	}

	// Build the final routes with precise typing
	public build(): ExtractDefinedRoutes<T> {
		return createEffectAPIRoutes(this.handlers, this.config);
	}
}
