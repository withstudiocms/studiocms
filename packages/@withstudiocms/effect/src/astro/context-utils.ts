import type { APIContext } from 'astro';
import { Effect, Schema } from '../effect.js';

/**
 * Reads and parses the JSON body from an API request within the given context.
 *
 * @template A - The expected type of the parsed JSON object.
 * @param context - An object containing the API request.
 * @returns An `Effect` that resolves with the parsed JSON object of type `A`, or fails with an `Error` if parsing fails.
 */
export const readAPIContextJson = <A>({ request }: APIContext): Effect.Effect<A, Error, never> =>
	Effect.tryPromise({
		try: () => request.json(),
		catch: () => new Error('Failed to parse JSON from Request'),
	});

/**
 * Reads and parses the JSON body from an API request within the given `APIContext`.
 * Optionally validates and decodes the parsed JSON using a provided `Schema`.
 *
 * @typeParam A - The expected type of the decoded JSON object.
 * @typeParam E - The error type that may be returned by the schema decoder.
 * @typeParam R - The environment type required by the schema decoder.
 * @param context - The API context containing the request to read from.
 * @param schema - Optional schema to validate and decode the parsed JSON.
 * @returns An `Effect` that yields the decoded JSON object of type `A`, or throws an error if parsing or decoding fails.
 */
export const parseAPIContextJson = <A, E, R>(
	context: APIContext,
	schema?: Schema.Schema<A, E, R>
): Effect.Effect<A, Error, R> =>
	Effect.gen(function* () {
		const json = yield* readAPIContextJson<A>(context);
		if (schema) return yield* Schema.decodeUnknown(schema)(json);
		return json as A;
	}).pipe(
		Effect.catchAll((error) => {
			console.error('Failed to read JSON:', error);
			return Effect.fail(new Error('Failed to read JSON'));
		})
	);

/**
 * Reads and parses form data from the API context's request object.
 *
 * Wraps the asynchronous `request.formData()` call in an `Effect.tryPromise`,
 * providing error handling if parsing fails.
 *
 * @param context - The API context containing the request object.
 * @returns An `Effect` that resolves to the parsed form data, or an error if parsing fails.
 */
export const readAPIContextFormData = ({
	request,
}: APIContext): Effect.Effect<FormData, Error, never> =>
	Effect.tryPromise({
		try: () => request.formData(),
		catch: () => new Error('Failed to parse formData from Request'),
	});

/**
 * Parses form data from an `APIContext` into an object, optionally validating and decoding it using a provided schema.
 *
 * @template A - The type of the parsed object.
 * @template E - The error type for schema decoding.
 * @template R - The environment type required by the effect.
 * @param context - The API context containing the form data to parse.
 * @param schema - Optional schema to validate and decode the form data.
 * @returns An `Effect` that yields the parsed object of type `A`, or fails with an `Error` if parsing or decoding fails.
 */
export const parseAPIContextFormDataToObject = <A, E, R>(
	context: APIContext,
	schema?: Schema.Schema<A, E, R>
): Effect.Effect<A, Error, R> =>
	Effect.gen(function* () {
		const formData = yield* readAPIContextFormData(context);
		const entries = Object.fromEntries(formData.entries());

		if (schema) return yield* Schema.decodeUnknown(schema)(entries);

		return entries as A;
	}).pipe(
		Effect.catchAll((error) => {
			console.error('Failed to read form data:', error);
			return Effect.fail(new Error('Failed to read form data'));
		})
	);

/**
 * Attempts to retrieve a string value from a FormData entry by key.
 * Returns an Effect that resolves to the string value if present and of type string,
 * or `null` if the entry is not a string.
 *
 * @param formData - The FormData object to extract the value from.
 * @param key - The key of the entry to retrieve.
 * @returns An Effect that resolves to the string value or `null`.
 */
export const parseFormDataEntryToString = (
	formData: FormData,
	key: string
): Effect.Effect<string | null, Error, never> =>
	Effect.try({
		try: () => {
			const value = formData.get(key);
			if (typeof value !== 'string') {
				return null;
			}
			return value;
		},
		catch: () => new Error(`Failed to parse FormData entry for key: ${key}`),
	});
