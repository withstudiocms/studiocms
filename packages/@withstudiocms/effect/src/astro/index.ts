export * from './api-route.js';
export * from './context-utils.js';
export * from './middleware.js';
export * from './response-helpers.js';
export * from './types.js';
