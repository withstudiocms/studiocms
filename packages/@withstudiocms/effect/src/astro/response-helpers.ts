import type { AllResponseOpts, createResponseOpts, OptionsResponseOpts } from './types.js';

/**
 * Generates a 204 No Content HTTP response for OPTIONS requests with appropriate CORS and allowed methods headers.
 *
 * @param opts - Configuration options for the response.
 * @param opts.allowedMethods - Array of HTTP methods allowed for the route (e.g., ['GET', 'POST']).
 * @param opts.allowedOrigins - Optional array of allowed origins for CORS. If not provided, defaults to '*'.
 * @param opts.headers - Optional additional headers to include in the response.
 * @returns A `Response` object configured for an OPTIONS request.
 */
export const OptionsResponse = (opts: OptionsResponseOpts): Response =>
	new Response(null, {
		status: 204,
		statusText: 'No Content',
		headers: {
			...opts.headers,
			Allow: `OPTIONS, ${opts.allowedMethods.join(', ')}`,
			'Access-Control-Allow-Origin': opts.allowedOrigins?.join(', ') || '*',
			Date: new Date().toUTCString(),
		},
	});

/**
 * Creates a `Response` object with a 405 "Method Not Allowed" status.
 *
 * @param opts - Options for configuring the response.
 * @param opts.allowedOrigins - An optional array of allowed origins for CORS. If not provided, defaults to '*'.
 * @param opts.headers - Optional additional headers to include in the response.
 * @returns A `Response` object with the specified headers and a 405 status.
 */
export const AllResponse = (opts?: AllResponseOpts): Response =>
	new Response(null, {
		status: 405,
		statusText: 'Method Not Allowed',
		headers: {
			...opts?.headers,
			'Access-Control-Allow-Origin': opts?.allowedOrigins?.join(', ') || '*',
			Date: new Date().toUTCString(),
		},
	});

/**
 * Creates a JSON HTTP response with customizable status, headers, and CORS settings.
 *
 * @template T - The type of the data to be serialized as JSON.
 * @param data - The data to be included in the response body.
 * @param opts - Optional settings for the response, including status code, status text, headers, and allowed origins.
 * @returns A `Response` object with the serialized JSON data and specified options.
 */
export const createJsonResponse = <T>(data: T, opts: createResponseOpts = {}) =>
	new Response(JSON.stringify(data), {
		status: opts.status || 200,
		statusText: opts.statusText || 'OK',
		headers: {
			'Content-Type': 'application/json',
			...opts?.headers,
			'Access-Control-Allow-Origin': opts?.allowedOrigins?.join(', ') || '*',
			Date: new Date().toUTCString(),
		},
	});

/**
 * Creates a `Response` object with a plain text body and customizable options.
 *
 * @param data - The string content to be sent in the response body.
 * @param opts - Optional settings for the response, including status, statusText, headers, and allowed origins.
 * @returns A `Response` object with the specified text content and options.
 */
export const createTextResponse = (data: string, opts: createResponseOpts = {}) =>
	new Response(data, {
		status: opts.status || 200,
		statusText: opts.statusText || 'OK',
		headers: {
			'Content-Type': 'text/plain',
			...opts?.headers,
			'Access-Control-Allow-Origin': opts?.allowedOrigins?.join(', ') || '*',
			Date: new Date().toUTCString(),
		},
	});

/**
 * Creates an HTML response with customizable status, headers, and allowed origins.
 *
 * @param data - The HTML string to be sent in the response body.
 * @param opts - Optional settings for the response, including status, statusText, headers, and allowedOrigins.
 * @returns A `Response` object with the specified HTML content and headers.
 */
export const createHtmlResponse = (data: string, opts: createResponseOpts = {}) =>
	new Response(data, {
		status: opts.status || 200,
		statusText: opts.statusText || 'OK',
		headers: {
			'Content-Type': 'text/html',
			...opts?.headers,
			'Access-Control-Allow-Origin': opts?.allowedOrigins?.join(', ') || '*',
			Date: new Date().toUTCString(),
		},
	});

/**
 * Creates a HTTP 302 redirect response with customizable headers.
 *
 * @param url - The URL to redirect to, set in the `Location` header.
 * @param opts - Optional settings for the response.
 * @param opts.headers - Additional headers to include in the response.
 * @param opts.allowedOrigins - Array of allowed origins for CORS, joined and set in the `Access-Control-Allow-Origin` header. Defaults to `*` if not provided.
 * @returns A `Response` object configured for a 302 redirect.
 */
export const createRedirectResponse = (url: string, opts: createResponseOpts = {}) =>
	new Response(null, {
		status: 302,
		statusText: 'Found',
		headers: {
			Location: url,
			...opts?.headers,
			'Access-Control-Allow-Origin': opts?.allowedOrigins?.join(', ') || '*',
			Date: new Date().toUTCString(),
		},
	});
