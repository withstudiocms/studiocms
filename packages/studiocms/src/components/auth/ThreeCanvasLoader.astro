---
import type { SiteConfigCacheObject } from 'studiocms:sdk/types';

interface Props {
	config: SiteConfigCacheObject;
}

// Get the site config
const { config } = Astro.props as Props;

const { loginPageBackground, loginPageCustomImage } = config.data;
---
<div 
    style="display: none;" 
    id="auth-pages-config" 
    data-config_background={loginPageBackground} 
    data-config_custom_image={loginPageCustomImage}
    />

<script>
import "studiocms:auth/scripts/three";
</script>