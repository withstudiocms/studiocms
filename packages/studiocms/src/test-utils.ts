import type { AstroIntegration, AstroIntegrationLogger } from 'astro';
import type { StudioCMSPlugin } from './plugins.js';
import type {
	SCMSAuthServiceFnOpts,
	SCMSDashboardFnOpts,
	SCMSFrontendFnOpts,
	SCMSImageServiceFnOpts,
	SCMSRenderingFnOpts,
	SCMSSiteMapFnOpts,
} from './schemas/index.js';

type HookRun<T> = { hasHook: boolean; hookResults: T };
export interface PluginHookResults {
	astroConfig: HookRun<{ integrations: AstroIntegration[] }>;
	studiocmsConfig: HookRun<{
		authService: Partial<SCMSAuthServiceFnOpts>;
		dashboard: Partial<SCMSDashboardFnOpts>;
		frontend: Partial<SCMSFrontendFnOpts>;
		imageService: Partial<SCMSImageServiceFnOpts>;
		rendering: Partial<SCMSRenderingFnOpts>;
		sitemap: Partial<SCMSSiteMapFnOpts>;
	}>;
}
/**
 * Utility class for testing StudioCMS plugins by simulating hook execution and collecting results.
 *
 * The `StudioCMSPluginTester` provides methods to:
 * - Create mock loggers for use in plugin hooks.
 * - Execute the `studiocms:astro:config` and `studiocms:config:setup` hooks, if present, on a given plugin.
 * - Collect and return integration and configuration data set by these hooks.
 * - Retrieve basic plugin metadata.
 *
 * @remarks
 * This class is intended for use in test environments to facilitate inspection and validation of plugin behavior.
 *
 * @example
 * ```typescript
 * const tester = new StudioCMSPluginTester(myPlugin);
 * const info = tester.getPluginInfo();
 * const hookResults = await tester.getHookResults();
 * ```
 */
export class StudioCMSPluginTester {
	private plugin: StudioCMSPlugin;
	private readonly injectedLogger?: AstroIntegrationLogger;

	constructor(plugin: StudioCMSPlugin, logger?: AstroIntegrationLogger) {
		this.plugin = plugin;
		this.injectedLogger = logger;
	}

	/**
	 * Creates a mock logger object for testing purposes.
	 *
	 * The returned logger implements the `AstroIntegrationLogger` interface and provides
	 * stubbed logging methods (`info`, `warn`, `error`, `debug`) that delegate to the corresponding
	 * `console` methods. It also includes a `fork` method, which returns a new mock logger
	 * instance with an attached label for testing logger forking behavior.
	 *
	 * @returns {AstroIntegrationLogger} A mock logger instance suitable for use in tests.
	 */
	private createMockLogger(): AstroIntegrationLogger {
		if (this.injectedLogger) return this.injectedLogger;
		// biome-ignore lint/suspicious/noExplicitAny: this is fine
		const logger: Record<string, any> = {
			info: console.log,
			warn: console.warn,
			error: console.error,
			debug: console.debug,
			fork: (label: string) => {
				// Each fork returns a new mock logger with the label attached for testing
				const forked = this.createMockLogger();
				forked.label = label;
				return forked;
			},
		};
		return logger as unknown as AstroIntegrationLogger;
	}

	/**
	 * Executes the 'studiocms:astro:config' hook if it exists on the plugin,
	 * providing a mock logger and a method to collect Astro integrations.
	 *
	 * @returns A promise that resolves to an object containing the collected integrations.
	 *
	 * @remarks
	 * This method checks if the plugin defines a 'studiocms:astro:config' hook as a function.
	 * If so, it invokes the hook with a mock logger and an `addIntegrations` callback,
	 * which accumulates integrations into an array. The collected integrations are then
	 * returned in an object.
	 */
	private async runAstroConfigHook() {
		const integrations: AstroIntegration[] = [];
		if (
			this.plugin.hooks['studiocms:astro:config'] &&
			typeof this.plugin.hooks['studiocms:astro:config'] === 'function'
		) {
			await this.plugin.hooks['studiocms:astro:config']({
				logger: this.createMockLogger(),
				addIntegrations: (newIntegrations) => {
					const toAdd = Array.isArray(newIntegrations) ? newIntegrations : [newIntegrations];
					integrations.push(...toAdd);
				},
			});
		}
		return { integrations };
	}

	/**
	 * Executes the `studiocms:config:setup` hook on the current plugin, if available,
	 * and collects configuration options set by the hook into partial option objects.
	 *
	 * This method mocks the hook's context by providing setter functions for various
	 * StudioCMS configuration aspects, such as authentication, dashboard, frontend,
	 * image service, rendering, and sitemap. Each setter captures the values provided
	 * by the hook and stores them in corresponding partial option objects.
	 *
	 * @returns A promise that resolves to an object containing the collected configuration
	 *          options for authentication service, dashboard, frontend, image service,
	 *          rendering, and sitemap.
	 *
	 * @remarks
	 * This utility is primarily intended for testing purposes, allowing inspection of
	 * configuration values set by a plugin's `studiocms:config:setup` hook.
	 */
	private async runStudioCMSConfigHook() {
		const authService: Partial<SCMSAuthServiceFnOpts> = {};
		const dashboard: Partial<SCMSDashboardFnOpts> = {};
		const frontend: Partial<SCMSFrontendFnOpts> = {};
		const imageService: Partial<SCMSImageServiceFnOpts> = {};
		const rendering: Partial<SCMSRenderingFnOpts> = {};
		const sitemap: Partial<SCMSSiteMapFnOpts> = {};

		if (
			this.plugin.hooks['studiocms:config:setup'] &&
			typeof this.plugin.hooks['studiocms:config:setup'] === 'function'
		) {
			await this.plugin.hooks['studiocms:config:setup']({
				logger: this.createMockLogger(),
				setAuthService: ({ oAuthProvider }) => {
					if (oAuthProvider) {
						authService.oAuthProvider = oAuthProvider;
					}
				},
				setDashboard: ({ dashboardGridItems, dashboardPages }) => {
					if (dashboardGridItems) {
						dashboard.dashboardGridItems = dashboardGridItems;
					}
					if (dashboardPages) {
						dashboard.dashboardPages = dashboardPages;
					}
				},
				setFrontend: ({ frontendNavigationLinks }) => {
					if (frontendNavigationLinks) {
						frontend.frontendNavigationLinks = frontendNavigationLinks;
					}
				},
				setImageService: ({ imageService: imgService }) => {
					if (imgService) {
						imageService.imageService = imgService;
					}
				},
				setRendering: ({ pageTypes }) => {
					if (pageTypes) {
						rendering.pageTypes = pageTypes;
					}
				},
				setSitemap: ({ sitemaps, triggerSitemap }) => {
					if (sitemaps) {
						sitemap.sitemaps = sitemaps;
					}
					sitemap.triggerSitemap = triggerSitemap;
				},
			});
		}

		return {
			authService,
			dashboard,
			frontend,
			imageService,
			rendering,
			sitemap,
		};
	}

	/**
	 * Retrieves information about the current plugin.
	 *
	 * @returns An object containing the plugin's identifier, name, minimum required StudioCMS version, and dependencies.
	 */
	public getPluginInfo() {
		return {
			identifier: this.plugin.identifier,
			name: this.plugin.name,
			studiocmsMinimumVersion: this.plugin.studiocmsMinimumVersion,
			requires: this.plugin.requires,
		};
	}

	/**
	 * Asynchronously retrieves the results of configured plugin hooks.
	 *
	 * @returns An object containing the presence and results of the 'studiocms:astro:config' and 'studiocms:config:setup' hooks.
	 * - `astroConfig`: Indicates if the 'studiocms:astro:config' hook exists and provides its execution results.
	 * - `studiocmsConfig`: Indicates if the 'studiocms:config:setup' hook exists and provides its execution results.
	 */
	public async getHookResults(): Promise<PluginHookResults> {
		return {
			astroConfig: {
				hasHook: !!this.plugin.hooks['studiocms:astro:config'],
				hookResults: await this.runAstroConfigHook(),
			},
			studiocmsConfig: {
				hasHook: !!this.plugin.hooks['studiocms:config:setup'],
				hookResults: await this.runStudioCMSConfigHook(),
			},
		};
	}
}
