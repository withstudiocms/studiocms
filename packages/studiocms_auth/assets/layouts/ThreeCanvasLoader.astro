---
// @ts-nocheck
import studioCMS_SDK_Cache from 'studiocms:sdk/cache';

// Get the site config
const {
	data: { loginPageBackground, loginPageCustomImage },
} = await studioCMS_SDK_Cache.GET.siteConfig();
---
<div 
    style="display: none;" 
    id="auth-pages-config" 
    data-config_background={loginPageBackground} 
    data-config_custom_image={loginPageCustomImage}
    />

<script>
import "studiocms:auth/scripts/three";
</script>