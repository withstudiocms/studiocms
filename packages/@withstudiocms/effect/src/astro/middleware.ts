import type { MiddlewareHandler } from 'astro';
import { defineMiddleware as _defineMiddleware } from 'astro/middleware';
import { runEffect } from '../index.js';
import type { EffectMiddlewareHandler, EffectMiddlewareRouterEntry } from './types.js';
import { buildMiddlewareSequence } from './utils/middleware.js';

/* v8 ignore next 10 */
/**
 * Defines an effect middleware by wrapping the provided handler function.
 * The middleware handler is executed within the `runEffect` context, allowing
 * for effectful operations to be managed and composed.
 *
 * @param fn - The effect middleware handler function to be wrapped.
 * @returns A middleware handler that executes the effect middleware within `runEffect`.
 */
export const defineMiddleware = (fn: EffectMiddlewareHandler): MiddlewareHandler =>
	_defineMiddleware(async (context, next) => await runEffect(fn(context, next)));

/* v8 ignore next 12 */
/**
 * Defines a middleware router that filters and applies middleware handlers based on path matching.
 *
 * The router is an array of route definitions, each specifying `includePaths` and `excludePaths`
 * for matching against the current request's pathname. Handlers whose paths match the criteria
 * are executed in sequence.
 *
 * @param router - An array of middleware route definitions, each containing path matching options and a handler.
 * @returns A composed middleware handler that executes all matching handlers in order, or calls `next` if none match.
 */
export const defineMiddlewareRouter = (router: EffectMiddlewareRouterEntry[]): MiddlewareHandler =>
	_defineMiddleware((context, next) => buildMiddlewareSequence(context, next, router));
