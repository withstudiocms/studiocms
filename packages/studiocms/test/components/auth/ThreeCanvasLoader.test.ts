/// <reference types="astro/client" />
import { experimental_AstroContainer as AstroContainer } from 'astro/container';
import { describe, expect, test } from 'vitest';
import ThreeCanvasLoader from '../../../src/components/auth/ThreeCanvasLoader.astro';
import { cleanAstroAttributes, MockAstroLocals } from '../../test-utils';

describe('ThreeCanvasLoader component', () => {
	test('render component', async () => {
		const container = await AstroContainer.create();
		const result = await container.renderToString(ThreeCanvasLoader, { locals: MockAstroLocals() });
		const cleanResult = cleanAstroAttributes(result, '/mock/path/ThreeCanvasLoader.astro');
		expect(cleanResult).toMatchSnapshot();
	});
});
