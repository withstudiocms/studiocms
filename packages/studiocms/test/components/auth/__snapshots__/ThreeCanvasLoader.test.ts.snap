// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ThreeCanvasLoader component > render component 1`] = `"<div style="display: none;" id="auth-pages-config" data-config_background="studiocms-curves"></div> <script type="module" src="/mock/path/ThreeCanvasLoader.astro?astro&type=script&index=0&lang.ts"></script>"`;
